### PR TITLE
DCP-3297: Added authentication to download  files from data hub

### PR DESCRIPTION
**Pull request by Manishemblebi**

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ena/backend/service/AccessionDetailsService.java [hunk 1]
```diff
@@ -140,38 +138,72 @@ public List<List<FileDetail>> fetchFileDetails(DownloadFormatEnum format,
...
+
         if (totalFiles > 0) {
             console.info("Downloading {} files in total", totalFiles);
         }
         return listList;
     }
 
+    private Set<String> getMissingAccessionsFromDataHub(DownloadJob downloadJob, List<List<FileDetail>> list) {
+
+        Set<String> userAccessions = new HashSet<>();
```
Comment (swatisuman1990) [inline src/main/java/uk/ac/ebi/ena/backend/service/AccessionDetailsService.java:160]:
```suggestion
        Set<String> userAccessions = new HashSet<>(downloadJob.getAccessionList());
```

File: src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java [hunk 2]
```diff
@@ -287,16 +292,26 @@ public List<EnaPortalResponse> getPortalResponses(List<String> accessionList, Do
...
+        log.info("portalAPIEndpoint: " + portalAPIEndpoint);
         String body = "includeAccessions=" + includeAccs;
         HttpHeaders httpHeaders = new HttpHeaders();
-        httpHeaders.add("Content-Type", URLENCODED);
-        httpHeaders.add("Accept", APPLICATION_JSON);
+        httpHeaders.add("Content-Type", Constants.URLENCODED);
+        httpHeaders.add("Accept", Constants.APPLICATION_JSON);
+
+        if (Objects.nonNull(authenticationDetail) && authenticationDetail.isAuthenticated()) {
+            System.out.println("SessionId : " + authenticationDetail.getSessionId());
```
Comment (swatisuman1990) [inline src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java:309]:
```suggestion
            console.info("SessionId : " + authenticationDetail.getSessionId());
```

File: src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java [hunk 3]
```diff
@@ -335,4 +356,50 @@ public void sendEmail(String recipientEmail, String message, String subject, Str
...
+
+            String portalAPIAuthEndpoint = Constants.PORTAL_API_EP + "/auth?dataPortal=" +
+                    CommonUtils.getDataPortalId(userName);
+            log.info("portalAPIAuthEndpoint: " + portalAPIAuthEndpoint);
+
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.add("Accept", Constants.APPLICATION_JSON);
+            httpHeaders.setBasicAuth(userName, password);
+
+            restTemplate.getInterceptors().add(new BasicAuthorizationInterceptor(userName, password));
```
Comment (swatisuman1990) [inline src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java:375]: BasicAuthorizationInterceptor is deprecated use BasicAuthenticationInterceptor instead

File: src/main/java/uk/ac/ebi/ena/backend/service/BackendServiceImpl.java [hunk 4]
```diff
@@ -45,20 +46,22 @@ public class BackendServiceImpl implements BackendService {
...
 
+
     @Override
     public void startDownload(DownloadFormatEnum format, String location, DownloadJob downloadJob,
-                              ProtocolEnum protocol, String asperaConnectLocation, String emailId) {
+                              ProtocolEnum protocol, String asperaConnectLocation, String emailId,
+                              AuthenticationDetail authenticationDetail) {
 
         console.info("Starting download for format:{} at download location:{},protocol:{}, asperaLoc:{}, emailId:{}",
                 format, location, protocol, asperaConnectLocation, emailId);
```
Comment (swatisuman1990) [inline src/main/java/uk/ac/ebi/ena/backend/service/BackendServiceImpl.java:56]: add dccUserName in the logs

**Review by swatisuman1990 — CHANGES_REQUESTED** [review]

**Review by swatisuman1990 — COMMENTED** [review]
Also need to add username and password in FileUtils.createDownloadScript so that script is created with credentials

File: README.md [hunk 5]
```diff
@@ -32,15 +32,20 @@ in the end will be prompted with the below options :
...
+* `--accessions` Comma separated list of accessions or file path to the accession list. If providing a list, it should
+  be a plain text file in TSV (tab separated values) format. If there are more than one column, the first column must be
+  the accessions. Header row is optional and will be ignored. Values can be enclosed in double quotes or not.
 * `--format` The format for the download (`eg : READS_FASTQ,READS_SUBMITTED,ANALYSIS_SUBMITTED,ANALYSIS_GENERATED`)
 * `--location` The location for the download
 * `--protocol` The protocol to be used for download.(`eg : FTP, ASPERA`). Default is FTP.
 * `--asperaLocation` The location of local Aspera Connect/CLI folder. Required if Protocol is Aspera.
 * `--email` The email at which one wishes to receive the alert.(optional)
+* `--dccUsername` Data hub Username .(Required only If you want to download the data from data hub)
```
[the remark this answers is not in the document]
Comment (Manishemblebi) [reply README.md:48]: 'Data hub Username' is correct message to display. https://www.ebi.ac.uk/ena/pathogens/explore

File: README.md [hunk 6]
```diff
@@ -32,15 +32,20 @@ in the end will be prompted with the below options :
...
+* `--dccUsername` Data hub password.(Required only If you want to download the data from data hub)
```
[the remark this answers is not in the document]
Comment (Manishemblebi) [reply README.md:49]: same.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (swatisuman1990) [reply README.md:48]: Here we are talking about the arguments for the script file and not the message.

File: src/main/java/uk/ac/ebi/ena/app/MainRunner.java [hunk 7]
```diff
@@ -94,31 +106,53 @@ public void run(String... args) {
...
+                                } else if (!"FTP".equals(protocol.name())) {
+                                    System.out.println("Only FTP protocol is supported to download the files from data hub");
+                                    log.error("Only FTP protocol is supported to download the files from data hub. Provided protocol is ", protocol);
+                                    throw new IllegalArgumentException();
+                                }
+                                authenticationDetail.setUserName(userName);
+                                authenticationDetail.setPassword(password);
+                                //Validate username and password
+                                if (!enaPortalService.authenticateUser(authenticationDetail)) {
+                                    log.error("Data hub username and or password are not correct.");
```
[the remark this answers is not in the document]
Comment (Manishemblebi) [reply src/main/java/uk/ac/ebi/ena/app/MainRunner.java]: Added "Data hub username and/or password is incorrect". As both can be incorrect.

File: src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java (hunk 3 above)
Comment (Manishemblebi) [reply src/main/java/uk/ac/ebi/ena/backend/service/EnaPortalService.java]: Removed BasicAuthorizationInterceptor and now passing basicAuth param to header.

**Review by swatisuman1990 — CHANGES_REQUESTED** [review]
Changes looks good. 
Just fix the failing test cases which is preventing the jar creation at the moment.

**Comment by Manishemblebi** [comment]
> Changes looks good. Just fix the failing test cases which is preventing the jar creation at the moment.

Fixed. Thanks

**Review by swatisuman1990 — APPROVED** [review]

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 image: openjdk:8
 
 variables:
-  APP_VERSION: "1.0.6"
+  APP_VERSION: "1.0.7"
```
Comment (suranjayathilaka) [inline .gitlab-ci.yml:4]: this can be v1.1.0 as we've added new functionality

File: README.md [hunk 9]
```diff
@@ -32,15 +32,20 @@ in the end will be prompted with the below options :
...
+* `--accessions` Comma separated list of accessions or file path to the accession list. If providing a list, it should
+  be a plain text file in TSV (tab separated values) format. If there are more than one column, the first column must be
+  the accessions. Header row is optional and will be ignored. Values can be enclosed in double quotes or not.
 * `--format` The format for the download (`eg : READS_FASTQ,READS_SUBMITTED,ANALYSIS_SUBMITTED,ANALYSIS_GENERATED`)
 * `--location` The location for the download
 * `--protocol` The protocol to be used for download.(`eg : FTP, ASPERA`). Default is FTP.
 * `--asperaLocation` The location of local Aspera Connect/CLI folder. Required if Protocol is Aspera.
 * `--email` The email at which one wishes to receive the alert.(optional)
+* `--dataHubUsername` Data hub username. (Required only If you want to download the data from data hub)
```
Comment (suranjayathilaka) [inline README.md:48]: from a data hub (dcc)

File: src/main/java/uk/ac/ebi/ena/app/MainRunner.java [hunk 10]
```diff
@@ -94,31 +106,54 @@ public void run(String... args) {
...
                     } else {
                         if (accessions != null && new File(downloadLocation).exists() && new File(downloadLocation).canWrite()) {
                             if (protocol == ProtocolEnum.ASPERA) {
                                 Assert.notNull(asperaLocation, ASPERA_PATH_MSG);
                             }
+                            //Download the data from dataHub
+                            if (StringUtils.isNotBlank(userName)) {
+                                authenticationDetail = new AuthenticationDetail();
+                                if (!StringUtils.startsWith(userName, "dcc_")) {
+                                    System.out.println("Please use data hub name (dcc username)");
```
Comment (suranjayathilaka) [inline src/main/java/uk/ac/ebi/ena/app/MainRunner.java:121]: Please provide a valid data hub name (dcc_username)

File: src/main/java/uk/ac/ebi/ena/app/MainRunner.java [hunk 11]
```diff
@@ -94,31 +106,54 @@ public void run(String... args) {
...
+                                    log.error("Invalid data hub name {} (dcc user) provided. ", userName);
+                                    throw new IllegalArgumentException("Invalid data hub name");
+                                } else if (!"FTP".equals(protocol.name())) {
+                                    System.out.println("Only FTP protocol is supported to download the files from data hub");
```
Comment (suranjayathilaka) [inline src/main/java/uk/ac/ebi/ena/app/MainRunner.java:125]: from a data hub

**Pull request merged**